dvc/repo/__init__.py

```diff
@@ -376,9 +387,10 @@ def stages(self, from_directory=None, check_dag=True):
         stages = []
         outs = []
 
-        ignore_file_handler = DvcIgnoreFileHandler(self.tree)
+        self.dvcignore.load_upper_levels(from_directory)
```

Comment:
This `.load_upper_levels()` is problematic:
- if it's called twice then you will have dup ignores, slowing things down
- if you forget to call it you will get a tricky to track bug

There should not be such thing at all. Possible strategies:
1. Walk entire repo dir on `dvcignore` object construction and build full `dvcignore`, never update it.
2. Same, but move `dvcignore` to cached property to make it lazy/not run unless needed.
3. Automatically call `._load_upper_levels()` on access, protect from reruns/dups.